android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt

```diff
@@ -780,6 +780,16 @@ abstract class EngineSession(
      */
     abstract fun toggleDesktopMode(enable: Boolean, reload: Boolean = false)
 
+    /**
+     * Checks if there is a rule for handling a cookie banner for the current website in the session.
+     *
+     * @param onSuccess callback invoked if the engine API returned a valid response. Please note
+     * that the response can be null - which can indicate a bug, a miscommunication
+     * or other unexpected failure.
```

Comment:
>      * @param onSuccess callback invoked if the engine API returned a valid response. Please note
>      * that the response can be null - which can indicate a bug, a miscommunication
>      * or other unexpected failure.
If `null` indicates `"indicates a bug, a miscommunication or other unexpected failure"` should we report it as an error instead (e.g `onError`)?  This will make the API concise and consistent for consumers.

Reply:
I don't think so because `onError` is called for the API error. I am not sure how likely it is that we return null, tbh

Reply:
We can file a bug for improving the error handling in AC, and improve all the APIs (`checkForFormData()`) etc. I am not sure if the pattern that is already being used in the codebase should be holding up this patch

Reply:
> Please note that the **response** can be null - which can indicate a bug, a miscommunication or other unexpected failure.

I think this looks like an implementation detail that shouldn't be pass to consumers, as they are not aware of a **response** is, this is a low detail on the `GeckoEngineSession` layer. Instead of us handling the GeckoView `null` response on our end, we are passing the handling to Fenix, Focus and all the apps that use AC.

Reply:
It's not an implementation detail :) It's one possible return value. Which I'm not sure is likely at all.

Reply:
Currently, we have `(Boolean?) -> Unit` - which makes it obvious that the parameter can be null, and so null needs to be handled. The caller can handle that in any way they want - they can display an error, they can have some default value, they can call the API again etc.

The variant `(Throwable?) -> Unit` implies we can have a nullable error there. Which raises a question, if there is no error, why are we calling `onError`? And please note that the null still needs to be handled, but in this case we have less options - because we are in an error handler, so the only way to handle this now is as if it is an error.

Can you explain to me why you think it's more convenient?

Reply:
Even if we call `onError` with a certain kind of exception like invalid data exception - that is still less options to handle the null

Reply:
For example, I'm seeing it from the perspective of the developer that will be consuming the API.

**Using the actual approach:** 
```kotlin
val onSuccess: (Boolean?) -> Unit = { 
    if (it == null){
        showErrorMessage() 
     } else {
        doSomethingWithTheValue(it)
     }
}

val onError: (Throwable)  -> Unit = { 
   logger(it)
   showErrorMessage() 
}

session.hasCookieBannerRuleForSession(onSuccess, onError)
```

**Use handling the `null` `response` and exposing a  `onSuccess: (Boolean) -> Unit`**

```kotlin
val onSuccess: (Boolean) -> Unit = { 
   doSomethingWithTheValue(it)
}

val onError: (Throwable)  -> Unit = { 
   logger(it)
   showErrorMessage() 
}

session.hasCookieBannerRuleForSession(onSuccess, onError)
```

We are not indicating enough information with  `onSuccess: (Boolean?) -> Unit` for consumers to handle the `null` diffidently than an error, it looks like we are making the API less consistent without providing any additional value to the consumer.

Reply:
Your suggestion is this though, isn't it?

```
val onError: (Throwable?)  -> Unit = { 
   logger.error(it)
   showErrorMessage() 
}
```

Or are you suggesting we do this in `then`: 

```
geckoResult.then(
  {value -> 
      if (value == null) onError(IllegalArgumentException)
     else doSomethingWith(value) 
   },
)
```

BTW, which exception are you suggesting?

Reply:
Addressed via zoom!